libcontainer/process_linux.go

```diff
@@ -420,7 +420,24 @@ func (p *initProcess) terminate() error {
 	if _, werr := p.wait(); err == nil {
 		err = werr
 	}
-	return err
+	return ignoreTerminateErrors(err)
+}
+
+// ignoreTerminateErrors returns nil if the given err matches an error known
+// to indicate that the terminate occurred successfully or err was nil, otherwise
+// err is returned unaltered.
+func ignoreTerminateErrors(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	// TODO(steve): Update these to none string checks if the runtime exports them.
+	switch err.Error() {
+	case "os: process already finished", "exec: Wait was already called":
```

Comment:
> We've requested the process be terminated, reporting an error due to the fact that the process had already gone is something the caller simply isn't interested in, as the desired effect has been achieved.

That depends on your perspective, and is not logic that should exist in a library. If you've asked us to destroy a container, under the assumption the container already exists then pretending as though it already exists can cause logic bugs and other such issues.

> Its analogous to a method to remove a file reporting "file doesn't exist", the caller simply doesn't care.

Some callers might care, and that's why `unlink(2)` (for example) **does** return `-ENOENT` in such cases. It's up to the user of the API to make a decision about what they care about -- not the library.

> 4. The cleanup code in container start() ensures everything is cleaned up by calling terminate.
> 5. terminate returns the bogus error os: process already finished which is then logged.

Right, so it looks like our code is incorrectly handling this case. The solution is to fix the calling code -- not to filter code inside this code.